### DISCUSSION TITLE
How to fix video fullscreening to window instead of whole display

### DISCUSSION THREAD

**Discussion by tywil04** [wailsapp/wails#2776] [Q&A]
As the title suggests, I am having issues with a video element. I am using the svelte template and this is the code for the video element:
```svelte
<video controls>
    <source src="https://interactive-examples.mdn.mozilla.net/media/cc0-videos/flower.webm" type="video/webm"/>
    <source src="https://interactive-examples.mdn.mozilla.net/media/cc0-videos/flower.mp4" type="video/mp4"/>
    <track kind="captions"/>
</video>
```

When I go to fullscreen the video element, instead of the expected behavior (fullscreening to the whole display) the element fullscreens to the wails window. Am I doing something wrong or is this a potential bug?

**Comment by leaanthony** [comment]
It's not a Wails big per se as you aren't using any Wails features here. It might be an issue with the webview though. What OS? Did you try the other OSes?

**Reply by tywil04** [reply]
I am using Windows 11, I just tested the same code on Linux (Fedora 38) and had the same issue. 

My full code is at: [https://github.com/tywil04/invidious-desktop](https://github.com/tywil04/invidious-desktop). I don't see any reason as to why the video element wont fullscreen correctly. My code is a lot more than the starter template so I will try and create a minimal reproducible example.

**Reply by tywil04** [reply]
I just tried creating a brand new wails project using the svelte template and appending the video element to the bottom of the page (I haven't touched anything else) and again, the video element only fullscreens to the window instead of the whole display. I don't know if its an issue with wails or something else in the chain.

**Reply by leaanthony** [reply]
Sounds like a limitation of the webview, which I guess would make sense as it has no control over the window. You could always trigger a fullscreen/unfullscreen when the button was pressed if that's an option?

**Reply by tywil04** [reply]
I originally didn't try that because I thought that it wouldn't work, now I have gotten it to work with this code:
```svelte
<script>
    import { WindowFullscreen, WindowUnfullscreen } from "../../wailsjs/runtime/runtime"

    const onFullscreen = (event) => {
        if (document.fullscreenElement === event.target) {
            WindowFullscreen()
        } else {
            WindowUnfullscreen()
        }
    }
</script>

// register event on document so all elements that fullscreen take up the whole display
<svelte:document on:fullscreenchange={onFullscreen}/>

<video controls>
    <source src="https://interactive-examples.mdn.mozilla.net/media/cc0-videos/flower.webm" type="video/webm"/>
    <source src="https://interactive-examples.mdn.mozilla.net/media/cc0-videos/flower.mp4" type="video/mp4"/>
    <track kind="captions"/>
</video>
```

The only issue with this method is that the fullscreen looks janky. I feel like it should just work out of the box.

**Reply by leaanthony** [reply]
I'm not sure I'd expect that from an embedded component TBH. This issue on the webview2feedback repo suggests you're doing the right thing: https://github.com/MicrosoftEdge/WebView2Feedback/issues/1834. We might be able to do something with the `CoreWebView2.ContainsFullScreenElementChanged` event handler... Please open a ticket for this 👍

**Reply by leaanthony** [reply]
I've done a draft PR for this here: https://github.com/wailsapp/wails/pull/2782. I'd be keen to know if that "jankiness" is the same as you're seeing.

**Reply by tywil04** [reply]
Yes, the video shows exactly the same effect I am seeing.

**Reply by leaanthony** [reply]
It looks like webview2 resizes the element first which makes it go to the left before the resize happens. They only raise the event after the fact.

**Reply by tywil04** [reply]
do you think there is any way to modify the behaviour so it looks better while fullscreening?

**Reply by leaanthony** [reply]
https://github.com/MicrosoftEdge/WebView2Feedback/issues/1834#issuecomment-1642779104

**Reply by tywil04** [reply]
that's cool, in the mean time, I have come up with a better looking solution that I have implemented in my app just using regular JavaScript. (I'm certain something similar could easily be implemented directly into wails, if you wanted).

The JavaScript I am using to get a better effect fullscreen effect:
```javascript

const onFullscreen = (event) => {
    if (document.fullscreenElement === event.target) {
        WindowMaximise()
        setTimeout(WindowFullscreen, 1)
    } else {
        WindowUnfullscreen()
        setTimeout(WindowUnmaximise, 1)
    }
}

document.addEventListener("fullscreenchange", onFullscreen)
```

Here's a video of the effect (I personally think it looks better, but it isn't perfect):
https://github.com/wailsapp/wails/assets/68185249/e1f8d4ef-0205-4320-ac7b-6e5732683fa6

**Reply by leaanthony** [reply]
Wow, that's significantly better than before!

**Reply by leaanthony** [reply]
@tywil04 - Would it be possible to get a video of the "before" - without the JS fix? I'm going to open a new ticket and it would be good to show the difference that small delay makes.

**Reply by tywil04** [reply]
Sure thing!

I have re-recorded the video with the delay fix so its using a minimal starting point instead of a test view in my project.

With delay fix:
https://github.com/wailsapp/wails/assets/68185249/c483381b-9636-41ff-af93-2db1faaeefe8

Without delay fix:
https://github.com/wailsapp/wails/assets/68185249/4b549601-333e-4d44-9563-a1ad350349a7

Without the small delay (1ms) the effect works when fullscreening but when exiting fullscreen the window won't unmaximise, its a really small quirk that I could only solve by introducing the small delay. I did also install the delay when fullscreening even though its not strictly necessary because the effect it produces looks more pleasing to me because it more closely matches the effect when exiting fullscreen.

**Reply by leaanthony** [reply]
Thanks so much for this 🙏I'd be so curious to know if this is an issue on Mac. I'll open the issue soon

**Reply by tywil04** [reply]
Unfortunately I don't have access to a Mac.
